lib/ansible/plugins/connection/ssh.py

```diff
@@ -44,6 +48,119 @@
 SSHPASS_AVAILABLE = None
 
 
+class AnsibleSshConnectionFailure(AnsibleConnectionFailure):
+    ''' the ssh connection plugin had a fatal error '''
+    pass
+
+
+# Could track retry attempts here. A little odd.
+class AnsibleAttemptedSshConnectionFailure(AnsibleConnectionFailure):
+    ''' the ssh connection plugin had a fatal error but may attempt to retry '''
+    pass
+
+class AnsibleSshInitialWriteConnectionFailure(AnsibleSshConnectionFailure):
+    ''' the ssh connection plugin had a fatal error during the initial data write '''
+    pass
+
+
+class AnsibleSshCommandErrorConnectionFailure(AnsibleSshConnectionFailure):
```

Comment:
I kind of debated if it made sense to share 'ssh connection exceptions' across paramiko/ssh.  In a lot of ways, there are some potential failures modes that are more conceptually 'ssh protocol' related than specific implementations (a failure in the crypto handshake setting up the transport layer for example).

The reasons I went for connection plugin specific exceptions were:

  * A shared exception would need to live in either:
       * base ansible (lib/ansible/errors/)
       * connection base (plugin/connection/__init__ for ex)   - either of these seems like it's pushing ssh specific concepts when it seems like there has been significant effort to remove ssh'isms from the app/lib
       * something new (plugin/connection/base_ssh.py ?)  - maybe makes more sense but that approach doesnt seem common for plugins  [plugins/strategy and plugins/shell/fish.py and plugins/action/net_template do however]
  * While something like an auth or transport failure is a general ssh concept, in practice the causes of the failures (and the potential remedies) tend to be implementation specific. Or tied to the implementation specific configuration or setup. 

Given those constraints, I pushed the more specific exceptions into the connection plugins.

It may make sense to push some slightly more specific base connection exceptions into lib/ansible/errors.py.
Other connection failure modes that exist across multiple connection types.

(psuedo-code sketch)
``` python
class AnsibleConnectionTransportFailure(AnsibleConnectionFailure):
    pass

class AnsibleConnectionAuthenticationFailure(AnsibleConnectionFailure):
   pass

# not a great name, but ssh channels, pipes/sockets (to dockerd  or popen comm for local for ex), 
class AnsibleConnectionChannelFailure(AnsibleConnectionFailure):
  pass

# Also not great names here, but something to separate failures for:
#  'the local command I run to make the connection failed'                  #  ssh, docker
#     and 
# 'the command I have to to run on the remote end as part of the connection failed'    #  "jail connection requires dd command in the jail" ,  no python, no powershell
class AnsibleConnectionCommandFailure(AnsibleConnectionFailure):
  pass

class AnsibleCommandConnectionFailure(AnsibleConnectionFailure):
  pass

etc
```